# IMPORT ALL REQUIRED LIBRARIES AND LOAD DATASET

## Required Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Load Dataset

In [3]:
nadp = pd.read_csv("C:\\Users\\saina\\Desktop\\DS_ML_AI\\Scaler\\Projects_or_Case_Studies_GIT\\Network_Anomaly_Detection_Project\\Datasets\\Network_anomaly_data.csv")

# BASIC EXPLORATORY DATA ANALYSIS

## First 5 rows

In [4]:
nadp.head()

duration protocoltype   service flag  srcbytes  dstbytes  land  \
0         0          tcp  ftp_data   SF       491         0     0   
1         0          udp     other   SF       146         0     0   
2         0          tcp   private   S0         0         0     0   
3         0          tcp      http   SF       232      8153     0   
4         0          tcp      http   SF       199       420     0   

   wrongfragment  urgent  hot  ...  dsthostsamesrvrate  dsthostdiffsrvrate  \
0              0       0    0  ...                0.17                0.03   
1              0       0    0  ...                0.00                0.60   
2              0       0    0  ...                0.10                0.05   
3              0       0    0  ...                1.00                0.00   
4              0       0    0  ...                1.00                0.00   

   dsthostsamesrcportrate  dsthostsrvdiffhostrate  dsthostserrorrate  \
0                    0.17                    0.00               0.00   
1                    0.88                    0.00               0.00   
2                    0.00                    0.00               1.00   
3                    0.03                    0.04               0.03   
4                    0.00                    0.00               0.00   

   dsthostsrvserrorrate  dsthostrerrorrate  dsthostsrvrerrorrate   attack  \
0                  0.00               0.05                  0.00   normal   
1                  0.00               0.00                  0.00   normal   
2                  1.00               0.00                  0.00  neptune   
3                  0.01               0.00                  0.01   normal   
4                  0.00               0.00                  0.00   normal   

   lastflag  
0        20  
1        15  
2        19  
3        21  
4        21  

[5 rows x 43 columns]

**Observations**
> 43 columns are present in the data

## Info

In [5]:
nadp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125973 entries, 0 to 125972
Data columns (total 43 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   duration                125973 non-null  int64  
 1   protocoltype            125973 non-null  object 
 2   service                 125973 non-null  object 
 3   flag                    125973 non-null  object 
 4   srcbytes                125973 non-null  int64  
 5   dstbytes                125973 non-null  int64  
 6   land                    125973 non-null  int64  
 7   wrongfragment           125973 non-null  int64  
 8   urgent                  125973 non-null  int64  
 9   hot                     125973 non-null  int64  
 10  numfailedlogins         125973 non-null  int64  
 11  loggedin                125973 non-null  int64  
 12  numcompromised          125973 non-null  int64  
 13  rootshell               125973 non-null  int64  
 14  suattempted         

**Observations**
> Out of 43 Columns, there are 15 float64 type columns, 24 int64 type columns, 4 object(String) type columns

> Dataset size is 41.3 MB

>   Total Number of rows are 125973.

>   It looks like there are no null values in any column. (Should check) 

## Null Value Detection

In [6]:
nadp.isna().sum()

duration                  0
protocoltype              0
service                   0
flag                      0
srcbytes                  0
dstbytes                  0
land                      0
wrongfragment             0
urgent                    0
hot                       0
numfailedlogins           0
loggedin                  0
numcompromised            0
rootshell                 0
suattempted               0
numroot                   0
numfilecreations          0
numshells                 0
numaccessfiles            0
numoutboundcmds           0
ishostlogin               0
isguestlogin              0
count                     0
srvcount                  0
serrorrate                0
srvserrorrate             0
rerrorrate                0
srvrerrorrate             0
samesrvrate               0
diffsrvrate               0
srvdiffhostrate           0
dsthostcount              0
dsthostsrvcount           0
dsthostsamesrvrate        0
dsthostdiffsrvrate        0
dsthostsamesrcportra

In [7]:
sum(nadp.isna().sum())

0

**Observations**
> There are no null values in the dataset. 

## Nunique of all columns

In [8]:
for i in nadp.columns:
    print(f"{i:<25} :   {nadp[i].nunique()}")

duration                  :   2981
protocoltype              :   3
service                   :   70
flag                      :   11
srcbytes                  :   3341
dstbytes                  :   9326
land                      :   2
wrongfragment             :   3
urgent                    :   4
hot                       :   28
numfailedlogins           :   6
loggedin                  :   2
numcompromised            :   88
rootshell                 :   2
suattempted               :   3
numroot                   :   82
numfilecreations          :   35
numshells                 :   3
numaccessfiles            :   10
numoutboundcmds           :   1
ishostlogin               :   2
isguestlogin              :   2
count                     :   512
srvcount                  :   509
serrorrate                :   89
srvserrorrate             :   86
rerrorrate                :   82
srvrerrorrate             :   62
samesrvrate               :   101
diffsrvrate               :   95
srvdiffhostra

**Observations**
> Most of the columns have int and float data types, so for most of the columns, the number of unique values are greater than 50

> Can check unique values for columns with nunique <= 50, to identify whether they are (continuous or discrete) and (nominal or ordinal)  

## Function to display all rows and columns

In [9]:
def display_all(df):     # For any Dataframe df
   with pd.option_context('display.max_rows',100): # Change number of rows accordingly
      with pd.option_context('display.max_columns',100): # Change number of columns accordingly
          display(df)

In [10]:
display_all(nadp.head())

duration protocoltype   service flag  srcbytes  dstbytes  land  \
0         0          tcp  ftp_data   SF       491         0     0   
1         0          udp     other   SF       146         0     0   
2         0          tcp   private   S0         0         0     0   
3         0          tcp      http   SF       232      8153     0   
4         0          tcp      http   SF       199       420     0   

   wrongfragment  urgent  hot  numfailedlogins  loggedin  numcompromised  \
0              0       0    0                0         0               0   
1              0       0    0                0         0               0   
2              0       0    0                0         0               0   
3              0       0    0                0         1               0   
4              0       0    0                0         1               0   

   rootshell  suattempted  numroot  numfilecreations  numshells  \
0          0            0        0                 0          0   
1          0            0        0                 0          0   
2          0            0        0                 0          0   
3          0            0        0                 0          0   
4          0            0        0                 0          0   

   numaccessfiles  numoutboundcmds  ishostlogin  isguestlogin  count  \
0               0                0            0             0      2   
1               0                0            0             0     13   
2               0                0            0             0    123   
3               0                0            0             0      5   
4               0                0            0             0     30   

   srvcount  serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  \
0         2         0.0            0.0         0.0            0.0   
1         1         0.0            0.0         0.0            0.0   
2         6         1.0            1.0         0.0            0.0   
3         5         0.2            0.2         0.0            0.0   
4        32         0.0            0.0         0.0            0.0   

   samesrvrate  diffsrvrate  srvdiffhostrate  dsthostcount  dsthostsrvcount  \
0         1.00         0.00             0.00           150               25   
1         0.08         0.15             0.00           255                1   
2         0.05         0.07             0.00           255               26   
3         1.00         0.00             0.00            30              255   
4         1.00         0.00             0.09           255              255   

   dsthostsamesrvrate  dsthostdiffsrvrate  dsthostsamesrcportrate  \
0                0.17                0.03                    0.17   
1                0.00                0.60                    0.88   
2                0.10                0.05                    0.00   
3                1.00                0.00                    0.03   
4                1.00                0.00                    0.00   

   dsthostsrvdiffhostrate  dsthostserrorrate  dsthostsrvserrorrate  \
0                    0.00               0.00                  0.00   
1                    0.00               0.00                  0.00   
2                    0.00               1.00                  1.00   
3                    0.04               0.03                  0.01   
4                    0.00               0.00                  0.00   

   dsthostrerrorrate  dsthostsrvrerrorrate   attack  lastflag  
0               0.05                  0.00   normal        20  
1               0.00                  0.00   normal        15  
2               0.00                  0.00  neptune        19  
3               0.00                  0.01   normal        21  
4               0.00                  0.00   normal        21

**Observations**
> Use display_all function whenever there is a need to print all the columns or rows.

## Unique values of columns with nunique <= 50

In [11]:
for i in nadp.columns:
    if nadp[i].nunique() <= 50:
        print(f"{i}")
        print(f"{nadp[i].unique()}")
        print()

protocoltype
['tcp' 'udp' 'icmp']

flag
['SF' 'S0' 'REJ' 'RSTR' 'SH' 'RSTO' 'S1' 'RSTOS0' 'S3' 'S2' 'OTH']

land
[0 1]

wrongfragment
[0 3 1]

urgent
[0 1 3 2]

hot
[ 0  5  6  4  2  1 28 30 22 24 14  3 15 25 19 18 77 17 11  7 20 12  9 10
  8 21 33 44]

numfailedlogins
[0 2 1 3 4 5]

loggedin
[0 1]

rootshell
[0 1]

suattempted
[0 1 2]

numfilecreations
[ 0  1  8  4  2 15 13 29 19 18  6 14  5 21 17 40  3 20 11 38 23 10 27 25
 12 16 28 26  7  9 33 22 43 36 34]

numshells
[0 1 2]

numaccessfiles
[0 1 2 3 5 4 8 6 7 9]

numoutboundcmds
[0]

ishostlogin
[0 1]

isguestlogin
[0 1]

attack
['normal' 'neptune' 'warezclient' 'ipsweep' 'portsweep' 'teardrop' 'nmap'
 'satan' 'smurf' 'pod' 'back' 'guess_passwd' 'ftp_write' 'multihop'
 'rootkit' 'buffer_overflow' 'imap' 'warezmaster' 'phf' 'land'
 'loadmodule' 'spy' 'perl']

lastflag
[20 15 19 21 18 17 16 12 14 11  2 13 10  9  8  7  3  5  1  6  0  4]



**Observation**
> **SHOULD COMMENT HERE AFTER UNDERSTANDING ABOUT EVERY COLUMN** 

## Descriptive statistics

In [12]:
display_all(nadp.describe())

duration      srcbytes      dstbytes           land  wrongfragment  \
count  125973.00000  1.259730e+05  1.259730e+05  125973.000000  125973.000000   
mean      287.14465  4.556674e+04  1.977911e+04       0.000198       0.022687   
std      2604.51531  5.870331e+06  4.021269e+06       0.014086       0.253530   
min         0.00000  0.000000e+00  0.000000e+00       0.000000       0.000000   
25%         0.00000  0.000000e+00  0.000000e+00       0.000000       0.000000   
50%         0.00000  4.400000e+01  0.000000e+00       0.000000       0.000000   
75%         0.00000  2.760000e+02  5.160000e+02       0.000000       0.000000   
max     42908.00000  1.379964e+09  1.309937e+09       1.000000       3.000000   

              urgent            hot  numfailedlogins       loggedin  \
count  125973.000000  125973.000000    125973.000000  125973.000000   
mean        0.000111       0.204409         0.001222       0.395736   
std         0.014366       2.149968         0.045239       0.489010   
min         0.000000       0.000000         0.000000       0.000000   
25%         0.000000       0.000000         0.000000       0.000000   
50%         0.000000       0.000000         0.000000       0.000000   
75%         0.000000       0.000000         0.000000       1.000000   
max         3.000000      77.000000         5.000000       1.000000   

       numcompromised      rootshell    suattempted        numroot  \
count   125973.000000  125973.000000  125973.000000  125973.000000   
mean         0.279250       0.001342       0.001103       0.302192   
std         23.942042       0.036603       0.045154      24.399618   
min          0.000000       0.000000       0.000000       0.000000   
25%          0.000000       0.000000       0.000000       0.000000   
50%          0.000000       0.000000       0.000000       0.000000   
75%          0.000000       0.000000       0.000000       0.000000   
max       7479.000000       1.000000       2.000000    7468.000000   

       numfilecreations      numshells  numaccessfiles  numoutboundcmds  \
count     125973.000000  125973.000000   125973.000000         125973.0   
mean           0.012669       0.000413        0.004096              0.0   
std            0.483935       0.022181        0.099370              0.0   
min            0.000000       0.000000        0.000000              0.0   
25%            0.000000       0.000000        0.000000              0.0   
50%            0.000000       0.000000        0.000000              0.0   
75%            0.000000       0.000000        0.000000              0.0   
max           43.000000       2.000000        9.000000              0.0   

         ishostlogin   isguestlogin          count       srvcount  \
count  125973.000000  125973.000000  125973.000000  125973.000000   
mean        0.000008       0.009423      84.107555      27.737888   
std         0.002817       0.096612     114.508607      72.635840   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       2.000000       2.000000   
50%         0.000000       0.000000      14.000000       8.000000   
75%         0.000000       0.000000     143.000000      18.000000   
max         1.000000       1.000000     511.000000     511.000000   

          serrorrate  srvserrorrate     rerrorrate  srvrerrorrate  \
count  125973.000000  125973.000000  125973.000000  125973.000000   
mean        0.284485       0.282485       0.119958       0.121183   
std         0.446456       0.447022       0.320436       0.323647   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

         samesrvrate    diffsrvrate  srvdiffhostrate   dsthostcount  \
count  125973.00000

In [13]:
display_all(nadp.describe(include="object"))

protocoltype service    flag  attack
count        125973  125973  125973  125973
unique            3      70      11      23
top             tcp    http      SF  normal
freq         102689   40338   74945   67343

**Observations**
> For most of the columns have min, 25th, 50th, 75th percentile as zeros. Only the max value is present. So those columns will have the many outliers in box plot.

## Duplicates Check

In [14]:
nadp.duplicated().sum()

np.int64(0)

In [16]:
nadp["duration"].describe()

count    125973.00000
mean        287.14465
std        2604.51531
min           0.00000
25%           0.00000
50%           0.00000
75%           0.00000
max       42908.00000
Name: duration, dtype: float64

In [17]:
nadp["service"].unique()

array(['ftp_data', 'other', 'private', 'http', 'remote_job', 'name',
       'netbios_ns', 'eco_i', 'mtp', 'telnet', 'finger', 'domain_u',
       'supdup', 'uucp_path', 'Z39_50', 'smtp', 'csnet_ns', 'uucp',
       'netbios_dgm', 'urp_i', 'auth', 'domain', 'ftp', 'bgp', 'ldap',
       'ecr_i', 'gopher', 'vmnet', 'systat', 'http_443', 'efs', 'whois',
       'imap4', 'iso_tsap', 'echo', 'klogin', 'link', 'sunrpc', 'login',
       'kshell', 'sql_net', 'time', 'hostnames', 'exec', 'ntp_u',
       'discard', 'nntp', 'courier', 'ctf', 'ssh', 'daytime', 'shell',
       'netstat', 'pop_3', 'nnsp', 'IRC', 'pop_2', 'printer', 'tim_i',
       'pm_dump', 'red_i', 'netbios_ssn', 'rje', 'X11', 'urh_i',
       'http_8001', 'aol', 'http_2784', 'tftp_u', 'harvest'], dtype=object)

In [18]:
nadp["srcbytes"].describe()

count    1.259730e+05
mean     4.556674e+04
std      5.870331e+06
min      0.000000e+00
25%      0.000000e+00
50%      4.400000e+01
75%      2.760000e+02
max      1.379964e+09
Name: srcbytes, dtype: float64

In [19]:
nadp["dstbytes"].describe()

count    1.259730e+05
mean     1.977911e+04
std      4.021269e+06
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      5.160000e+02
max      1.309937e+09
Name: dstbytes, dtype: float64

In [20]:
nadp["land"].value_counts()

land
0    125948
1        25
Name: count, dtype: int64

In [22]:
nadp["wrongfragment"].value_counts()

wrongfragment
0    124883
3       884
1       206
Name: count, dtype: int64

In [21]:
nadp.columns

Index(['duration', 'protocoltype', 'service', 'flag', 'srcbytes', 'dstbytes',
       'land', 'wrongfragment', 'urgent', 'hot', 'numfailedlogins', 'loggedin',
       'numcompromised', 'rootshell', 'suattempted', 'numroot',
       'numfilecreations', 'numshells', 'numaccessfiles', 'numoutboundcmds',
       'ishostlogin', 'isguestlogin', 'count', 'srvcount', 'serrorrate',
       'srvserrorrate', 'rerrorrate', 'srvrerrorrate', 'samesrvrate',
       'diffsrvrate', 'srvdiffhostrate', 'dsthostcount', 'dsthostsrvcount',
       'dsthostsamesrvrate', 'dsthostdiffsrvrate', 'dsthostsamesrcportrate',
       'dsthostsrvdiffhostrate', 'dsthostserrorrate', 'dsthostsrvserrorrate',
       'dsthostrerrorrate', 'dsthostsrvrerrorrate', 'attack', 'lastflag'],
      dtype='object')

In [41]:
for i in nadp.columns:
    print(i)
    print(nadp[i].value_counts())
    print()
    

duration
duration
0        115955
1          1989
2           843
3           557
4           351
          ...  
12039         1
9235          1
41254         1
36613         1
679           1
Name: count, Length: 2981, dtype: int64

protocoltype
protocoltype
tcp     102689
udp      14993
icmp      8291
Name: count, dtype: int64

service
service
http         40338
private      21853
domain_u      9043
smtp          7313
ftp_data      6860
             ...  
tftp_u           3
http_8001        2
aol              2
harvest          2
http_2784        1
Name: count, Length: 70, dtype: int64

flag
flag
SF        74945
S0        34851
REJ       11233
RSTR       2421
RSTO       1562
S1          365
SH          271
S2          127
RSTOS0      103
S3           49
OTH          46
Name: count, dtype: int64

srcbytes
srcbytes
0        49392
8         3691
1         2432
44        2334
45        2089
         ...  
2282         1
36696        1
2427         1
2633         1
13425        1
Name: c

In [27]:
nadp["hot"].value_counts()

hot
0     123302
2       1037
1        369
28       277
30       256
4        173
6        140
5         76
24        68
19        57
22        55
3         54
18        45
14        30
20         9
7          5
15         4
11         3
9          2
25         2
44         2
17         1
77         1
12         1
10         1
8          1
21         1
33         1
Name: count, dtype: int64

In [30]:
nadp["numcompromised"].unique()

array([   0,    3,    2,    1,   19,    5,    4,   38,  107,  462,  110,
          6,  151,    7,   31,    9,  157,   13,  558,   12,  371,  217,
         83,  193,  789,  520,  884,  457,  767,  756,   21,   10,  281,
         23,   44,  622,   18,   15,  716,  202,  121,  682,  537,  345,
         17,    8,  166, 7479,  307,   40,  405,  751,   54,  761,  102,
        237,  456,   22,  809,  198,   37,   75,  378,   16, 1739,  174,
         46,   74,  543,  691,   94,  676,  175,  373,   14,   78, 1043,
        177,  247,  538,  187,  407,  258,  568,  349,  452,  375,   11])

In [31]:
nadp["numcompromised"].value_counts()

numcompromised
0      124687
1         976
2          98
4          40
3          38
        ...  
568         1
349         1
452         1
375         1
11          1
Name: count, Length: 88, dtype: int64

In [37]:
nadp["numfilecreations"].value_counts()

numfilecreations
0     125686
1        151
2         41
4         13
8          5
15         5
5          5
17         5
3          5
10         5
11         4
12         4
7          4
18         4
40         3
25         3
14         3
20         3
6          3
26         3
9          2
23         2
13         2
21         1
29         1
19         1
27         1
28         1
16         1
38         1
33         1
22         1
43         1
36         1
34         1
Name: count, dtype: int64

In [39]:
nadp['count'].value_counts()

count
1      27763
2       9474
3       3962
4       3550
5       2980
       ...  
413        4
501        4
506        3
504        3
505        2
Name: count, Length: 512, dtype: int64

In [40]:
nadp['srvcount'].value_counts()

srvcount
1      25398
2      12820
3       6336
4       5526
5       4636
       ...  
430        2
503        1
501        1
413        1
498        1
Name: count, Length: 509, dtype: int64

In [28]:
nadp.columns

Index(['duration', 'protocoltype', 'service', 'flag', 'srcbytes', 'dstbytes',
       'land', 'wrongfragment', 'urgent', 'hot', 'numfailedlogins', 'loggedin',
       'numcompromised', 'rootshell', 'suattempted', 'numroot',
       'numfilecreations', 'numshells', 'numaccessfiles', 'numoutboundcmds',
       'ishostlogin', 'isguestlogin', 'count', 'srvcount', 'serrorrate',
       'srvserrorrate', 'rerrorrate', 'srvrerrorrate', 'samesrvrate',
       'diffsrvrate', 'srvdiffhostrate', 'dsthostcount', 'dsthostsrvcount',
       'dsthostsamesrvrate', 'dsthostdiffsrvrate', 'dsthostsamesrcportrate',
       'dsthostsrvdiffhostrate', 'dsthostserrorrate', 'dsthostsrvserrorrate',
       'dsthostrerrorrate', 'dsthostsrvrerrorrate', 'attack', 'lastflag'],
      dtype='object')

In [42]:
(nadp["samesrvrate"]+nadp["diffsrvrate"]).unique()

array([1.  , 0.23, 0.12, 0.22, 0.11, 0.2 , 0.14, 0.12, 0.08, 0.06, 0.22,
       0.15, 0.07, 0.09, 0.35, 1.5 , 0.13, 0.3 , 0.18, 0.17, 0.58, 1.01,
       0.24, 0.45, 0.17, 1.34, 0.1 , 0.16, 0.75, 0.07, 0.62, 0.25, 0.38,
       0.23, 0.06, 0.95, 0.36, 1.03, 0.21, 0.24, 0.11, 0.52, 0.15, 0.44,
       0.09, 0.72, 0.27, 0.33, 0.5 , 0.21, 0.34, 0.47, 0.19, 1.33, 0.43,
       0.57, 0.28, 0.6 , 0.33, 0.48, 0.32, 0.39, 0.51, 0.42, 0.37, 0.41,
       0.83, 0.3 , 0.59, 0.55, 1.12, 1.02, 0.31, 0.89, 0.1 , 0.93, 0.63,
       0.86, 1.04, 0.53, 0.61, 0.05, 0.42, 1.2 , 0.79, 1.13, 0.54, 0.35,
       0.68, 0.97, 0.77, 0.46, 0.84, 0.86, 0.86, 0.92, 0.45, 1.25, 0.96,
       0.26, 0.59, 0.34, 0.43, 0.84, 0.58, 0.56, 0.7 , 0.9 , 0.66, 0.94,
       0.68, 0.29, 0.6 , 0.8 , 0.41, 0.47, 0.4 , 0.39, 0.67, 0.64, 0.99,
       0.91, 0.93, 1.16, 1.11, 0.67, 0.62, 0.64, 0.82, 0.73, 1.05, 1.17,
       0.18, 1.08, 0.49, 0.87, 0.88, 0.69, 1.06, 0.44, 0.03, 0.71, 0.48,
       0.9 , 0.89, 0.92, 0.4 , 0.29, 0.  , 0.28, 0.

In [43]:
display_all(nadp.head())

duration protocoltype   service flag  srcbytes  dstbytes  land  \
0         0          tcp  ftp_data   SF       491         0     0   
1         0          udp     other   SF       146         0     0   
2         0          tcp   private   S0         0         0     0   
3         0          tcp      http   SF       232      8153     0   
4         0          tcp      http   SF       199       420     0   

   wrongfragment  urgent  hot  numfailedlogins  loggedin  numcompromised  \
0              0       0    0                0         0               0   
1              0       0    0                0         0               0   
2              0       0    0                0         0               0   
3              0       0    0                0         1               0   
4              0       0    0                0         1               0   

   rootshell  suattempted  numroot  numfilecreations  numshells  \
0          0            0        0                 0          0   
1          0            0        0                 0          0   
2          0            0        0                 0          0   
3          0            0        0                 0          0   
4          0            0        0                 0          0   

   numaccessfiles  numoutboundcmds  ishostlogin  isguestlogin  count  \
0               0                0            0             0      2   
1               0                0            0             0     13   
2               0                0            0             0    123   
3               0                0            0             0      5   
4               0                0            0             0     30   

   srvcount  serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  \
0         2         0.0            0.0         0.0            0.0   
1         1         0.0            0.0         0.0            0.0   
2         6         1.0            1.0         0.0            0.0   
3         5         0.2            0.2         0.0            0.0   
4        32         0.0            0.0         0.0            0.0   

   samesrvrate  diffsrvrate  srvdiffhostrate  dsthostcount  dsthostsrvcount  \
0         1.00         0.00             0.00           150               25   
1         0.08         0.15             0.00           255                1   
2         0.05         0.07             0.00           255               26   
3         1.00         0.00             0.00            30              255   
4         1.00         0.00             0.09           255              255   

   dsthostsamesrvrate  dsthostdiffsrvrate  dsthostsamesrcportrate  \
0                0.17                0.03                    0.17   
1                0.00                0.60                    0.88   
2                0.10                0.05                    0.00   
3                1.00                0.00                    0.03   
4                1.00                0.00                    0.00   

   dsthostsrvdiffhostrate  dsthostserrorrate  dsthostsrvserrorrate  \
0                    0.00               0.00                  0.00   
1                    0.00               0.00                  0.00   
2                    0.00               1.00                  1.00   
3                    0.04               0.03                  0.01   
4                    0.00               0.00                  0.00   

   dsthostrerrorrate  dsthostsrvrerrorrate   attack  lastflag  
0               0.05                  0.00   normal        20  
1               0.00                  0.00   normal        15  
2               0.00                  0.00  neptune        19  
3               0.00                  0.01   normal        21  
4               0.00                  0.00   normal        21